### PR TITLE
FIX: don't allow inviting more than `max_allowed_message_recipients` setting allows

### PR DESCRIPTION
**Pull request by OsamaSayegh**

### CODE REVIEW

**Comment by OsamaSayegh** [comment]
@eviltrout thanks for the feedback! There is a different guardian method `can_invite_to?` that checks for other things such as sufficient trust level etc. I see how it's confusing now, let me try to clear it up.

**Comment by OsamaSayegh** [comment]
@eviltrout let me know if it's clearer now, thanks.

File: lib/guardian.rb [hunk 1]
```diff
@@ -272,6 +272,7 @@ def can_invite_to?(object, groups = nil)
 
     if object.is_a?(Topic) && object.private_message?
       return false unless SiteSetting.enable_personal_messages?
+      return false if object.reached_recipients_limit? && !is_staff?
```
Comment (tgxworld) [inline lib/guardian.rb]: I think we should update the specs to reflect this change.

File: app/models/topic_user.rb [hunk 2]
```diff
@@ -40,7 +40,7 @@ def notification_reasons
     def auto_notification(user_id, topic_id, reason, notification_level)
       should_change = TopicUser
         .where(user_id: user_id, topic_id: topic_id)
-        .where("notifications_reason_id IS NULL OR (notification_level < :min AND notification_level > :max)", min: notification_level, max: notification_levels[:regular])
+        .where("notifications_reason_id IS NULL OR (notification_level < :max AND notification_level > :min)", max: notification_level, min: notification_levels[:regular])
```
Comment (tgxworld) [inline app/models/topic_user.rb]: Is this change intentional because the code path that has been updated doesn't seem to be affected by this change here.

File: app/controllers/topics_controller.rb [hunk 3]
```diff
@@ -538,6 +538,12 @@ def invite
       group_names: params[:group_names]
     )
 
+    if !guardian.is_staff? && topic.reached_recipients_limit?
```
Comment (tgxworld) [inline app/controllers/topics_controller.rb]: I feel like the responsibility of checking if there are too many recipients should be part of `Topic#invite` instead.

Comment (OsamaSayegh) [reply app/controllers/topics_controller.rb]: I did initially put the code in the model method, but then I wanted to show a message to users when they try to go above the recipients limit and here is where I got a problem. We have `guardian.ensure_can_invite_to!(..)` called in the controller `invite` method, and this guardian method does check if there are too many recipients, so when it fails it's going to raise a generic 403 error. Is it OK if I repeat the code in the model method? I can't think of a cleaner way to achieve that.

File: app/models/topic_user.rb (hunk 2 above)
Comment (OsamaSayegh) [reply app/models/topic_user.rb]: Yeah it's not related but I was reading the code and found the placeholders names very confusing. It's just mathematically impossible to find a value larger than `max` and smaller than `min` at the same time.

**Comment by OsamaSayegh** [comment]
I've made a couple changes here:

1. "Automatically track topics I enter..." and "When I post in a topic, set that topic to..." user preferences will no longer apply to PMs

2. groups will now take up 1 slot in PM

File: app/controllers/topics_controller.rb [hunk 4]
```diff
@@ -538,6 +544,12 @@ def invite
       group_names: params[:group_names]
     )
 
+    unless pm_has_slots?(topic)
```
Comment (SamSaffron) [inline app/controllers/topics_controller.rb]: This does not feel right cause you are allowed it appears to invite multiple groups, so you can cheat and invite 30 groups ? no?

Comment (OsamaSayegh) [reply app/controllers/topics_controller.rb]: Yes that will allow you to cheat and invite 30 groups. Maybe we should have another site setting for max groups in a PM? Because making every group member take up a slot doesn't feel quite right to me.

Comment (SamSaffron) [reply app/controllers/topics_controller.rb]: I think this is fine... just add `pm_has_slots?(topic, groups.length)` or something like that.

Comment (OsamaSayegh) [reply app/controllers/topics_controller.rb]: the `groups` variable here represents groups that you want invited user to be a member of, when you invite by email. These groups won't be invited to the message.

Inviting a group to a message is handled via a separate controller method `invite_group`, and you can't invite groups in bulk, only one by one.

We do a check like this
`topic_allowed_users.count + topic_allowed_groups.count >= SiteSetting.max_allowed_message_recipients`
when you invite a group or user.

File: lib/post_creator.rb [hunk 5]
```diff
@@ -531,7 +531,8 @@ def track_topic
     if @user.staged
       TopicUser.auto_notification_for_staging(@user.id, @topic.id, TopicUser.notification_reasons[:auto_watch])
     else
-      notification_level = @user.user_option.notification_level_when_replying || NotificationLevels.topic_levels[:tracking]
+      notification_level = TopicUser.where(user: @user, topic: @topic).pluck(:notification_level).first if @topic.private_message?
```
[the remark this answers is not in the document]
Comment (OsamaSayegh) [reply lib/post_creator.rb]: Agreed, we can just `return` if topic is PM.

File: app/models/topic_user.rb (hunk 2 above)
Comment (tgxworld) [reply app/models/topic_user.rb]: @OsamaSayegh I wonder if this might be a on-demand fix for a bug that was introduced in the past. Maybe have a look through the git commit history?

**Comment by SamSaffron** [comment]
cool lets try this out!
